### DISCUSSION TITLE
Reducing CSS selector nesting and making base class names longer

### DISCUSSION THREAD

**Discussion by mayank99** [iTwin/iTwinUI#55] [Ideas]
We have a stylelint rule for limiting selector nesting.

https://github.com/iTwin/iTwinUI/blob/df726cad40fb0600647f28077c6a17acca485825/.stylelintrc.json#L10

And because of the way we conventionally write css in this repo, we often find ourselves hitting that limit (in header, in sidenav, in slider, and in so many other places).

The other problem we run into is with clashing selectors (e.g. we use the same classes in completely different ways in different components and they end up conflicting).

I think we should start making our selectors flatter by making base classnames (not modifiers) longer where we need to.

---

[Inception Rule](https://thesassway.com/beginner-the-inception-rule/):
> _you shouldn’t be too specific or mimicking the DOM at any point. If you find yourself more than four levels deep, that’s a red flag. Of course there are times when you’ll be forced to go there, but it’s not something you should be doing too much._

Other good reads:
- [Beware Selector Nesting](https://www.sitepoint.com/beware-selector-nesting-sass/) by Kitty Giraudel on Sitepoint
- [CSS tricks style guide](https://css-tricks.com/sass-style-guide/#maximum-nesting-three-levels-deep) by Chris Coyier
- [Avoid nested selectors](https://thesassway.com/avoid-nested-selectors-for-more-modular-css/) on TheSassWay

**Comment by mayank99** [comment]
Here's an example of a problem that comes from mimicking DOM structure: https://github.com/iTwin/iTwinUI-react/issues/30#issuecomment-828399142

We want to wrap a button with span but that breaks button-group styling because of `>`.
https://github.com/iTwin/iTwinUI/blob/53065e4aaf2ee08d38f676885851c9c7db3c36d8/src/button/button-group.scss#L8

**Reply by mayank99** [reply]
Another example: https://github.com/iTwin/iTwinUI-react/pull/45#discussion_r631998227

All because of `>` and nesting
https://github.com/iTwin/iTwinUI/blob/9e66a3f6cc84fcb42ad8b1c97ebfc7b2228e6dc5/src/header/header.scss#L34

**Reply by mayank99** [reply]
Another one: #123

And another tangentially related one: #125 (comes from too many pseudo-selectors, not nesting)

**Reply by mayank99** [reply]
another example of the CSS required to achieve a custom hover state

```scss
.iui-windows-close:hover,
.iui-windows-close:focus {
  @include themed {
    background-color: t(iui-color-background-negative) !important;
  }
  .iui-icon {
    @include themed {
      fill: t(iui-color-foreground-accessory) !important;
    }
  }
}
```

without `!important`:
![image](https://user-images.githubusercontent.com/9084735/136830768-c35c4e88-8110-42a7-bc94-411a48cefb33.png)

**Comment by maxime4000** [comment]
While I agree that limiting nesting has some benefits, I wouldn't impose a lint about it because some component can be complex and nesting help understanding how the component is designed. I would said, be aware that nesting can be bad and try to avoid that, but imposing limit is not helpful.

**Comment by bentleyvk** [comment]
One of the most popular component lib says to recommends to use longer names and avoid scoping: https://github.com/palantir/blueprint/wiki/Coding-guidelines#css-selectors

**Reply by mayank99** [reply]
They link to [css tips from Mozilla](https://hacks.mozilla.org/2016/05/css-coding-techniques/), which has this really good line:
> _do not let nesting generate CSS rules you wouldn’t type yourself_

And look at our generated selector 😄 
https://github.com/iTwin/iTwinUI/blob/d7e09b4ae4b9b9979081f18235511a583c38dab2/lib/css/all.css#L683
